_studio/shared/umc/codec/mpeg2_dec/src/umc_mpeg2_frame.cpp

```diff
@@ -0,0 +1,316 @@
+
+    MPEG2DecoderFrameInfo::~MPEG2DecoderFrameInfo()
+    {
+        std::for_each(slices.begin(), slices.end(),
+            std::default_delete<MPEG2Slice>()
+        );
+    }
+
+    void MPEG2DecoderFrameInfo::Reset()
+    {
+        isField       = false;
+        isBottomField = false;
+        frameType     = (FrameType)0;
+        isFilled      = false;
+
+        std::for_each(slices.begin(), slices.end(),
+            std::default_delete<MPEG2Slice>()
+        );
+
+        slices.resize(0);
+        sliceCount = 0;
+        FreeReferenceFrames();
+
+    }
+
+    void MPEG2DecoderFrameInfo::AddSlice(MPEG2Slice* slice)
+    {
+        if (0 == sliceCount) // on the first slice
+        {
+            const auto pic    = slice->GetPicHeader();
+            const auto picExt = slice->GetPicExtHeader();
+            isField       =  picExt.picture_structure != FRM_PICTURE;
+            isBottomField = (picExt.picture_structure == BOTTOM_FLD_PICTURE);
+            frameType     = (FrameType)pic.picture_coding_type;
+        }
+
+        slices.push_back(slice);
+        sliceCount++;
+    }
+
+    void MPEG2DecoderFrameInfo::UpdateReferences(const DPBType & dpb)
+    {
+        const auto pic    = GetSlice(0)->GetPicHeader();
+        const auto picExt = GetSlice(0)->GetPicExtHeader();
+
+        if (MPEG2_I_PICTURE == pic.picture_coding_type)
+            return;
+
+        // 7.6.3.5 Prediction in P-pictures (for IP pair)
+        if (FRM_PICTURE != frame.pictureStructure)
+        {
+            uint8_t currFieldIndex = frame.GetNumberByParity(picExt.picture_structure == BOTTOM_FLD_PICTURE);
+            if ((1 == currFieldIndex) && (MPEG2_P_PICTURE == pic.picture_coding_type)) // current picture is P and the 2nd field
+            {
+                if (MPEG2_I_PICTURE == frame.GetAU(0)->GetSlice(0)->GetPicHeader().picture_coding_type) // the first field is I
+                {
+                    references.forward = &frame;
+                    references.forward->IncrementReference();
+                    return;
+                }
+            }
+        }
+
+        // 7.6.1 Frame/Field prediction
+        DPBType refFrames = dpb;
+        refFrames.remove_if(
+            [this](MPEG2DecoderFrame const * f)
+            {
+                return (f == &frame) || !f->IsRef();
+            }
+        );
+
+        // Find latest ref frame
+        auto i = std::max_element(std::begin(refFrames), std::end(refFrames),
+            [](MPEG2DecoderFrame const* f1, MPEG2DecoderFrame const* f2)
+            {
+                return f1->decOrder < f2->decOrder;
+            }
+        );
+
+        auto ref = (i == std::end(refFrames)) ? nullptr : *i;
+
+        // If no reference frame is found, use 'self' reference.
+        // This is to handle cases without leading I picture
+        if (!ref)
+        {
+            ref = &frame;
+        }
+
+        if (MPEG2_P_PICTURE == pic.picture_coding_type)
+        {
+            if (ref)
```

Comment:
Can't find how is implemented for second field in P frame, which utilizes same polarity field from previous frame and first field of the current frame. (7.6.2.1 in 13818-2 spec)

Reply:
it's here. Just previous in dec order frame

Reply:
It is for "the first field is I"

Reply:
I don't understand what you're looking for. IP case is above. P/B frame are at 128 and below.

Reply:
I'm looking for 2nd P field case. Not necessary in IP, but in PP pair.
Second field refers to 2(!) frames - previous and current (first field of the current).

Reply:
I see what you mean. For HW implementation we need to pass a reference surface(s).  So reference from the first field is available to the driver without an extra information from decoder